File: src/main/java/cz/wa/secretagent/game/controller/GameMenuController.java
package cz.wa.secretagent.game.controller;

import org.lwjgl.input.Keyboard;
import org.springframework.beans.factory.annotation.Required;

import cz.wa.secretagent.menu.MenuKeys;
import cz.wa.secretagent.menu.window.GFrame;
import cz.wa.secretagent.menu.window.component.selectable.GSelectable;
import cz.wa.secretagent.simulation.GameController;
import cz.wa.secretagent.worldinfo.WorldHolder;

/**
 * Controls all menus. (move cursor, enter, back)
 * 
 * @author Ondrej Milenovsky
 */
public class GameMenuController implements GameController {

    private static final long serialVersionUID = -8439774283488072370L;

    private WorldHolder worldHolder;
    private transient boolean canActivate = false;

    @Override
    public void activate() {
        canActivate = false;
    }

    @Override
    public void processInput(double timeS) {
        MenuKeys menuKeys = worldHolder.getMenuHolder().getKeys();

        if (Keyboard.isKeyDown(menuKeys.kBack)) {
            if (canActivate) {
                GFrame frame = worldHolder.getMenuHolder().getTopFrame();
                frame.callClose();
                canActivate = false;
            }
        } else if (Keyboard.isKeyDown(menuKeys.kUp)) {
            if (canActivate) {
                GFrame frame = worldHolder.getMenuHolder().getTopFrame();
                frame.selectPrevComponent();
                canActivate = false;
            }
        } else if (Keyboard.isKeyDown(menuKeys.kDown)) {
            if (canActivate) {
                GFrame frame = worldHolder.getMenuHolder().getTopFrame();
                frame.selectNextComponent();
                canActivate = false;
            }
        } else if (Keyboard.isKeyDown(menuKeys.kEnter)) {
            if (canActivate) {
                GFrame frame = worldHolder.getMenuHolder().getTopFrame();
                GSelectable selected = frame.getSelectedComponent();
                if (selected != null) {
                    selected.activate();
                }
                canActivate = false;
            }
        } else if (Keyboard.isKeyDown(menuKeys.kLeft)) {
            if (canActivate) {
                GFrame frame = worldHolder.getMenuHolder().getTopFrame();
                GSelectable selected = frame.getSelectedComponent();
                if ((selected != null) && !selected.changeLeft()) {
                    frame.selectPrevComponent();
                }
                canActivate = false;
            }
        } else if (Keyboard.isKeyDown(menuKeys.kRight)) {
            if (canActivate) {
                GFrame frame = worldHolder.getMenuHolder().getTopFrame();
                GSelectable selected = frame.getSelectedComponent();
                if ((selected != null) && !selected.changeRight()) {
                    frame.selectNextComponent();
                }
                canActivate = false;
            }
        } else {
            canActivate = true;
        }
    }

    public WorldHolder getWorldHolder() {
        return worldHolder;
    }

    @Required
    public void setWorldHolder(WorldHolder worldHolder) {
        this.worldHolder = worldHolder;
    }

}

File: src/main/kotlin/secretAgent/game/controller/GameMenuController.kt
package secretAgent.game.controller

import cz.wa.secretagent.simulation.GameController
import cz.wa.secretagent.worldinfo.WorldHolder
import org.lwjgl.input.Keyboard
import org.springframework.beans.factory.annotation.Required

/**
 * Controls all menus. (move cursor, enter, back)
 *
 * @author Ondrej Milenovsky
 */
class GameMenuController : GameController {

    lateinit var worldHolder: WorldHolder
    @Transient
    private var canActivate = false

    override fun activate() {
        canActivate = false
    }

    override fun processInput(timeS: Double) {
        val menuKeys = worldHolder.menuHolder.keys!!

        when {
            Keyboard.isKeyDown(menuKeys.kBack) -> if (canActivate) {
                val frame = worldHolder.menuHolder.topFrame
                frame!!.callClose()
                canActivate = false
            }
            Keyboard.isKeyDown(menuKeys.kUp) -> if (canActivate) {
                val frame = worldHolder.menuHolder.topFrame
                frame!!.selectPrevComponent()
                canActivate = false
            }
            Keyboard.isKeyDown(menuKeys.kDown) -> if (canActivate) {
                val frame = worldHolder.menuHolder.topFrame
                frame!!.selectNextComponent()
                canActivate = false
            }
            Keyboard.isKeyDown(menuKeys.kEnter) -> if (canActivate) {
                val frame = worldHolder.menuHolder.topFrame
                val selected = frame!!.selectedComponent
                selected?.activate()
                canActivate = false
            }
            Keyboard.isKeyDown(menuKeys.kLeft) -> if (canActivate) {
                val frame = worldHolder.menuHolder.topFrame
                val selected = frame!!.selectedComponent
                if (selected?.changeLeft() == false)
                    frame.selectPrevComponent()
                canActivate = false
            }
            Keyboard.isKeyDown(menuKeys.kRight) -> if (canActivate) {
                val frame = worldHolder.menuHolder.topFrame
                val selected = frame!!.selectedComponent
                if (selected?.changeRight() == false)
                    frame.selectNextComponent()
                canActivate = false
            }
            else -> canActivate = true
        }
    }

    companion object {
        private const val serialVersionUID = -8439774283488072370L
    }
}